File: scripts/extract-image2fun.py
import subprocess, re, json, urllib.parse
from html import unescape
from urllib.parse import urlparse, parse_qs

BASE = 'https://image2.fun'
html = subprocess.check_output(['curl', '-sSL', BASE + '/'], text=True)

# Extract card slugs and images from main page
cards = []
seen = set()
for m in re.finditer(r'<a[^>]*href="(/p/[^"]+)"[^>]*>', html):
    slug = m.group(1)
    if slug in seen:
        continue
    seen.add(slug)
    block = html[m.start():min(len(html), m.start() + 3000)]

    # Image URL
    img_match = re.search(r'<img[^>]*src="([^"]+)"', block)
    img = img_match.group(1) if img_match else None
    if img and '_next/image' in img:
        try:
            parsed = urlparse(unescape(img))
            params = parse_qs(parsed.query)
            img = params.get('url', [img])[0]
            img = urllib.parse.unquote(img)
        except:
            pass

    # Title from alt or h3
    title_match = re.search(r'alt="([^"]{5,})"', block)
    title = title_match.group(1) if title_match else None
    if not title:
        title_match = re.search(r'<h3[^>]*>([^<]+)<', block)
        title = title_match.group(1) if title_match else None

    # Category badge
    cat_match = re.search(r'class="[^"]*inline-flex[^"]*rounded-md[^"]*text-secondary-foreground[^"]*"[^>]*>([^<]+)', block)
    category = cat_match.group(1).strip() if cat_match else None

    cards.append({'slug': slug, 'image': img, 'title': title, 'category': category})

print(f'Found {len(cards)} cards on main page')

# Fetch detail pages in batches to extract prompts
BATCH_SIZE = 40
results = []

for i, card in enumerate(cards[:BATCH_SIZE]):
    slug = card['slug']
    print(f'[{i+1}/{min(BATCH_SIZE,len(cards))}] {slug}')

    try:
        detail = subprocess.check_output(['curl', '-sSL', BASE + slug], text=True, timeout=30)

        # Look for prompt text in __next_f.push blocks
        # The RSC format uses \uXXXX encoding for Chinese chars
        prompt = None
        for push_m in re.finditer(r'__next_f\.push\(\[1,"(.*?)"\]\)', detail, re.DOTALL):
            raw = push_m.group(1)
            try:
                decoded = raw.encode().decode('unicode_escape')
            except:
                continue

            # Check if this block contains the prompt (long Chinese text)
            if re.search(r'[一-鿿]{30,}', decoded) and len(decoded) > 200:
                # Extract just the prompt text - it's the longest continuous Chinese block
                chinese_segments = re.findall(r'[一-鿿][^"]{50,1500}', decoded)
                for seg in chinese_segments:
                    # Clean up HTML entities and escaping
                    seg = seg.replace('\\n', '\n').replace('\\"', '"').replace('&quot;', '"')
                    seg = seg.replace('\\(', '(').replace('\\)', ')')
                    if len(seg) > 100 and not seg.startswith('children'):
                        prompt = seg
                        break
                if prompt:
                    break

        results.append({**card, 'prompt': prompt})
        if prompt:
            print(f'  prompt: {prompt[:120]}...')
        else:
            print(f'  WARNING: no prompt found')

    except Exception as e:
        print(f'  ERROR: {e}')
        results.append({**card, 'prompt': None})

    if i < min(BATCH_SIZE, len(cards)) - 1:
        import time
        time.sleep(0.5)

with open('/Users/wanghans/gpt-image2-templates/scripts/extracted_image2fun_prompts.json', 'w') as f:
    json.dump(results, f, ensure_ascii=False, indent=2)

with_prompt = sum(1 for r in results if r['prompt'])
print(f'\nDone. {with_prompt}/{len(results)} with prompts')
print('Saved extracted_image2fun_prompts.json')

File: scripts/extract-image2fun.mjs
import { writeFile } from 'node:fs/promises'

const BASE = 'https://image2.fun'

// ---- Step 1: fetch main page and extract card slugs + images ----
const html = await fetch(BASE).then(r => r.text())

const slugPattern = /href="(\/p\/[^"]+)"/g
const slugs = [...new Set([...html.matchAll(slugPattern)].map(m => m[1]))]
console.log(`Found ${slugs.length} card slugs`)

// Extract card data blocks
const cards = []
const seen = new Set()
const linkRe = /<a[^>]*href="(\/p\/[^"]+)"[^>]*>/g
let match
while ((match = linkRe.exec(html)) !== null) {
  const slug = match[1]
  if (seen.has(slug)) continue
  seen.add(slug)

  const block = html.slice(match.index, Math.min(html.length, match.index + 3000))

  // image src
  const imgRe = /<img[^>]*src="([^"]+)"/.exec(block)
  let img = imgRe ? imgRe[1] : null
  // unwrap _next/image query param
  if (img && img.includes('_next/image')) {
    try {
      const qs = new URL(img, BASE).searchParams
      img = decodeURIComponent(qs.get('url') || '')
    } catch {}
  }

  // title from alt or h3
  const titleRe = /alt="([^"]{5,})"/.exec(block)
  const h3Re = /<h3[^>]*>([^<]+)</.exec(block)
  const title = titleRe?.[1] || h3Re?.[1] || null

  cards.push({ slug, image: img, title })
}

console.log(`Extracted ${cards.length} cards`)
const first20 = cards.slice(0, 20)
for (const c of first20) {
  console.log(`\n  ${c.slug}`)
  console.log(`  title: ${c.title}`)
  console.log(`  image: ${(c.image || 'N/A').slice(0, 150)}`)
}

await writeFile(
  new URL('extracted_cards.json', import.meta.url),
  JSON.stringify(cards, null, 2)
)
console.log('\nSaved extracted_cards.json')

// ---- Step 2: fetch detail pages for prompts ----
const BATCH_SIZE = 30
const results = []

for (let i = 0; i < Math.min(BATCH_SIZE, cards.length); i++) {
  const card = cards[i]
  const slugPath = card.slug
  console.log(`\n[${i + 1}/${BATCH_SIZE}] fetching ${slugPath}...`)

  try {
    const detailHtml = await fetch(BASE + slugPath).then(r => r.text())

    // Extract prompt from __next_f.push data
    const promptRe = /__next_f\.push\(\[1,"[^"]*\\(?:n|")?([^"]*(?:生成|创建|设计|绘制|写实|风格|插画|色调|构图|光影)[^"]{30,3000})/g
    let promptMatch
    let prompt = null
    while ((promptMatch = promptRe.exec(detailHtml)) !== null) {
      const candidate = promptMatch[1]
        .replace(/\\n/g, '\n')
        .replace(/\\"/g, '"')
        .replace(/&quot;/g, '"')
      if (candidate.length > 100 && !prompt) {
        prompt = candidate
      }
    }

    // If no prompt found via the above pattern, try a broader search
    if (!prompt) {
      const broadRe = /__next_f\.push\(\[1,"[^"]*\\n([^"]{200,3000})[^"]*"\]\)/
      const broadMatch = broadRe.exec(detailHtml)
      if (broadMatch) {
        prompt = broadMatch[1].replace(/\\n/g, '\n').replace(/\\"/g, '"').replace(/&quot;/g, '"')
      }
    }

    results.push({
      ...card,
      prompt: prompt || null
    })

    if (prompt) {
      console.log(`  prompt: ${prompt.slice(0, 150)}...`)
    } else {
      console.log(`  WARNING: no prompt found`)
    }

  } catch (err) {
    console.error(`  ERROR: ${err.message}`)
    results.push({ ...card, prompt: null })
  }

  // Small delay to be polite
  if (i < Math.min(BATCH_SIZE, cards.length) - 1) {
    await new Promise(r => setTimeout(r, 500))
  }
}

await writeFile(
  new URL('extracted_with_prompts.json', import.meta.url),
  JSON.stringify(results, null, 2)
)

const withPrompt = results.filter(r => r.prompt)
console.log(`\nDone. ${withPrompt.length}/${results.length} have prompts`)
console.log('Saved extracted_with_prompts.json')
